### PR TITLE
sql, sql/parser, sql/sem/tree: sanitize/simplify the handling of names

### PR DESCRIPTION
**Pull request by knz**
Forked off #21753 for easier review.

Prior to this patch, the type `UnresolvedName` was an array of things
called "name parts", and could include stars, array indirections or
subscripts. The corresponding logic to normalize unresolved names into
table names, column names or other things was complicated.

In the light of a desired upcoming change to support both catalog and
schema specifiations in names, this code was welcoming a
simplification.

The simplification came from the realization that the syntactic
structure for a table or column path is unambiguous and the "item of
interest" is always in the last position. My previous (mistaken)
understanding was that a column name could be followed by other name
parts in the grammar; however, this is not valid SQL.

- to access an array column, the grammar rules to parse `a.b.c[d]`
  parse first `a.b.c` as a column item, then `<expr> [d]` as an array
  subscript expression. The `[d]` part is not part of the grammar for
  names, never has.

- to access a field in a column of a compound type (these are not
  supported by CockroachDB yet, but the grammar mirrors pg's which
  does support them), pg's SQL dialect has it that one *must* write
  `(a.b.c).field` with grouping parentheses; a direct path of
  `a.b.c.field` is not valid. (More precisely, `a.b.c.field`
  *always* refers to the column `field` of table `c` in schema `b` in
  catalog `a`, and the query would error out if there is no such
  table/column.)

  In short, an unresolved column name again always has a period-separated
  path with the column name in last position.

The predecessor to this patch already had simplified the grammar rules
to account for the known maximum length of a qualified name. This
patch now simplifies the AST and all uses to further capitalize on the
new understanding and simplify the normalization and pretty-printing
rules.

Release note: None

### CODE REVIEW

**Comment by knz** [comment]
Im sorry Jordan and Nathan you don't need to review this. I'm sure justin is just as qualified.

**Comment by justinj** [comment]
:lgtm:! Could `NumParts` be made into a method that gets computed based on the length of `Parts` and whether `Star` is set?

---

Review status: 0 of 20 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/config/zone.go, line 152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22269#-L4DDQ9kTeGJidJt3CLN:-L4DDQ9kTeGJidJt3CLO:b4k72xn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9606bb15590eb972df232f096ba68f296eddc617/pkg/config/zone.go#L152)):*
> ```Go
> 			TableNameReference: &tree.UnresolvedName{
> 				NumParts: 3,
> 				Parts:    tree.NameParts{string(zs.Partition), string(tn.TableName), string(tn.SchemaName)},
> ```

I feel like there are sufficiently many instances of this pattern that `UnresolvedName` might benefit from a constructor that computes the `NumParts` itself, like
```
TableNameReference: tree.NewUnresolvedName([]string{string(zs.Partition), string(tn.TableName), string(tn.SchemaName)}),
```
That's mostly an aesthetic thing though, don't think it makes much difference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22269#-:-L4DDLkkntZaD6uW-O3T:bcrvng7)*

**Comment by knz** [comment]
> gets computed based on the length of Parts and whether Star is set?

Well that's the point, `Parts` is now of type `[4]string` (that was the insight, which enables us to get rid of the heap allocations) so it doesn't have a "length" (or rather, its length is always 4). The information about which items in that array are actually specified is now carried by NumParts and is not available anywhere else. 

---

Review status: 0 of 20 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/config/zone.go, line 152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22269#-L4DDQ9kTeGJidJt3CLN:-L4DEO-j5clj-GgiHR_H:b6j1ovw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9606bb15590eb972df232f096ba68f296eddc617/pkg/config/zone.go#L152)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I feel like there are sufficiently many instances of this pattern that `UnresolvedName` might benefit from a constructor that computes the `NumParts` itself, like
```
TableNameReference: tree.NewUnresolvedName([]string{string(zs.Partition), string(tn.TableName), string(tn.SchemaName)}),
```
That's mostly an aesthetic thing though, don't think it makes much difference.
</blockquote></details>

Again doing so would introduce a heap allocation (for the `[]string{...}`) which is what this PR is trying to get rid of.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22269#-:-L4DDh-EY1-TCWcfAaao:b-9g1br2)*

**Comment by justinj** [comment]
Oh I see, makes sense. The allocation perspective wasn't mentioned above the struct or in the commit message, though that's compelling - could you add it? I also observe that the struct could contain a slice that's backed by its own array which might simplify some of the implementations here, though that would introduce more pointers for the GC to keep track of and might just complicate things.

**Comment by knz** [comment]
Thanks! Added a comment to clarify.

**Comment by jordanlewis** [comment]
:lgtm:

---

Reviewed 19 of 20 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22269#-:-L4HI-4Dsb95i8zCivQl:bnfp4nl)*
